### DISCUSSION TITLE
How to implement appsettings in Uno?

### DISCUSSION THREAD

**Discussion by TonyK-FmGroup** [unoplatform/uno#8541] [Q&A]
Hello,
What would be the most appropriate way to supply app settings to all heads in an uno solution? Can an appsettings.json be added to the shared project? Can individual settings then be overridden in some heads and not others? 
In addition, we'd need to have separate dev and production settings, naturally.

**Comment by pkar70** [comment]
Do you mean settings that are read-only, or settings that can be changed (but not persistent), or persistent settings with default values?
For second, you can use Microsoft.Extension.Configuration.

For third, I created a ConfigurationProviders for using with Microsoft.Extension.Configuration. And now, I have multiple levels:
* static (compile time) defaults
* static defaults for debug build (not for release build)
* environment variables, 
* Windows.Storage.ApplicationData.Current.RoamingSettings
* Windows.Storage.ApplicationData.Current.LocalSettings
* json file roaming
* json file local
* command line
:)

And both JSON and ApplicationData are read/write (and persistent).

**Reply by TonyK-FmGroup** [reply]
For my current needs, a combination of 1 and 3.  We need some read only settings that are defined at build time and then some user settings with defaults.

**Reply by pkar70** [reply]
You can use code from my lib (https://github.com/pkar70/MyLibs/tree/master/Vblib14), and for Windows.Storage.ApplicationData, https://github.com/pkar70/MyLibs/blob/master/UnoApp/UnoApp.Shared/pkarModule-Uno4-0-0.cs (#region ".Net configuration - UWP settings", lines 2347+, and 394+). 
I don't know how to convert it to Nuget :)

**Comment by jeromelaban** [comment]
What's the original intent for app settings and for which platform? Is it to configure the application ([environment variables](https://github.com/unoplatform/Uno.Wasm.Bootstrap#environment-variables) can be used) or to store [user settings](https://docs.microsoft.com/en-us/uwp/api/windows.storage.applicationdata.localsettings?view=winrt-22000) ?

**Reply by TonyK-FmGroup** [reply]
The very first requirement is really just to have certain strings in a single location instead of in code, for example the URL to certain APIs. Initially these won't change so I'm tempted to just use a resw (although that feels a bit dirty to do it that way), but at some time in the future we'll need to allow the user to change some of these settings, well, maybe :) 
I can see that I'm still trying to get my head around writing for multiple platforms and that, for example, we probably wouldn't implement 'settings' for WASM the same as we would for a desktop app, more likely they'd be in a user profile remotely.

**Reply by jeromelaban** [reply]
Indeed. Each platform behaves a bit differently with regards to compile-time settings. User settings though can be used the same way on all platforms.

**Reply by TonyK-FmGroup** [reply]
Thanks again, re the environment variables, we're trying that but finding that in local debug, the system env vars are not read, it will only read those in the csproj.  Is this by design?

the csproj has 
```
<ItemGroup>
		<WasmShellMonoEnvironment Include="APIBaseURL" Value="https://localhost:58775/" />
</ItemGroup>
```

Code reads with:
```
System.Environment.GetEnvironmentVariable("APIBaseURL");
```

I have a system environment variable with a matching name which is different to the csproj. The value read is always that from the csproj. If I remove if from the csproj, then the returned value is NULL.

**Reply by jeromelaban** [reply]
What is a "system environment variable" in your description? Is it a variable you're defining in your environment running `dotnet build` ?

**Reply by TonyK-FmGroup** [reply]
Sorry, I'm on windows, the system environment variables are set "System Properties" 

![image](https://user-images.githubusercontent.com/103487880/174724974-06596098-aae1-4598-9698-95d93d52dd4f.png)

**Reply by jeromelaban** [reply]
Got it. So you want to capture the environment variables values to include their values in the Wasm build. 

If so, you'll need to propagate that value to the `WasmShellMonoEnvironment` item group:

```xml
<PropertyGroup>
   <APIBaseURL Condition="'$(APIBaseURL)'==''">https://localhost:58775/</APIBaseUrl>
</PropertyGroup>
<ItemGroup>
   <WasmShellMonoEnvironment Include="APIBaseURL" Value="$(APIBaseURL)" />
</ItemGroup>
```

**Reply by TonyK-FmGroup** [reply]
A bit more than this, after the build and deploy, we want to consume the environment variables and have them override any in the csproj, just as would happen with an asp.net deployment.  
In short, what we need is that URL to be specified differently on each deployment target (dev,test,prod) as part of our pipeline.

**Reply by jeromelaban** [reply]
Got it. At this time, there is no explicit feature for this kind of behavior. You could try downloading a custom file at the root from the source site [using content files](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-additional-files.html#support-for-additional-content-files), then alter your app's behavior with it, or you could build an app package for each stage of your pipeline.

**Reply by TonyK-FmGroup** [reply]
This approach would mean we'd have to compile differently for each environment in our pipeline which is something we're trying to avoid.

**Reply by jeromelaban** [reply]
Building multiple times, yes, but not with the content file at the root approach I mentioned above.

This would be the equivalent of an appsettings.json that you can modify when deploying.

**Reply by TonyK-FmGroup** [reply]
Thank you, we ended up taking a slightly different route by ensuring that our UI and our API share a common Base URL (although different physical deployments) and thus we were able to us JS to read the current window's URL to use as the Base URL. This works well for us in our environment and we're one step closer to getting our POC out, although only to WASM due to limitations in the UWP deployment in Citrix.

If this wasn't possible, another suggestion we had was, because we deploy to containers, to use a fixed base url in the code and then using nginx to rewrite / redirect that in each environment.

**Comment by asadullah-refat-89** [comment]
Hello @jeromelaban . I have been trying to use the appsettings.json file in my project. Nothing fancy just wanna read a string from the file and use the IConfiguration interface. I have installed the Uno.Extensions.Configuration and all relevant Uno.Extensions NuGet packages. But it seems these two methods .UseAppSettings() and .UseConfiguration<AppInfo>() are not supported. I learned about these from this video:
https://platform.uno/blog/a-deep-dive-into-uno-extensions/
But it looks like I can not reference these methods.

**Reply by asadullah-refat-89** [reply]
I have also looked up the official docs for this. But these methods do not exist in the latest NuGet packages anymore.
I read this: https://platform.uno/docs/articles/external/uno.extensions/doc/Overview/Configuration/ConfigurationOverview.html

**Reply by sasakrsmanovic** [reply]
A ping to @nickrandolph here too.

**Reply by nickrandolph** [reply]
Hey my apologies with regards to docs - we're in process of updating them.
Best way to get an idea of how to use appsettings.json is to take a look at the unoapp-extensions dotnet new template (https://www.nuget.org/packages/Uno.Extensions.Templates)
Install:
dotnet new --install Uno.Extensions.Templates::2.3.0-dev.129
Create app:
dotnet new unoapp-extensions -o MyAppName
